### PR TITLE
Fix using and, or, xor in desul with MSVC

### PR DESCRIPTION
**Pull request by masterleinad**
Corresponds to kokkos/kokkos#7124.
[kokkos/kokkos/actions/runs/9845036970/job/27179696028?pr=7123](https://github.com/kokkos/kokkos/actions/runs/9845036970/job/27179696028?pr=7123) and [godbolt.org/z/5384bWqWY](https://godbolt.org/z/5384bWqWY) show that using and, or, and xor as macro arguments is problematic with MSVC when <ciso646> is included. This pull request makes sure that we post-fix and pre-fix these arguments instead.

### CODE REVIEW

**Comment by dalg24** [comment]
Why change the hip and sycl implementations?

**Comment by masterleinad** [comment]
> Why change the hip and sycl implementations?

Mostly for consistency. It's unclear if we couldn't encounter the same problem with a different configuration.

File: atomics/include/desul/atomics/Fetch_Op_CUDA.hpp [hunk 1]
```diff
@@ -69,56 +69,56 @@ inline __device__       unsigned int device_atomic_fetch_inc_mod(  unsigned int*
...
+DESUL_IMPL_CUDA_DEVICE_ATOMIC_FETCH_OP_INTEGRAL(fetch_sub)
 
-DESUL_IMPL_CUDA_DEVICE_ATOMIC_FETCH_OP_INTEGRAL(inc)
-DESUL_IMPL_CUDA_DEVICE_ATOMIC_FETCH_OP_INTEGRAL(dec)
+DESUL_IMPL_CUDA_DEVICE_ATOMIC_FETCH_OP_INTEGRAL(fetch_inc)
+DESUL_IMPL_CUDA_DEVICE_ATOMIC_FETCH_OP_INTEGRAL(fetch_dec)
 
-DESUL_IMPL_CUDA_DEVICE_ATOMIC_FETCH_OP(inc_mod, unsigned int)
-DESUL_IMPL_CUDA_DEVICE_ATOMIC_FETCH_OP(dec_mod, unsigned int)
+DESUL_IMPL_CUDA_DEVICE_ATOMIC_FETCH_OP(_inc_mod, unsigned int)
```
Comment (dalg24) [inline atomics/include/desul/atomics/Fetch_Op_CUDA.hpp:120]: This looks wrong

File: atomics/include/desul/atomics/Fetch_Op_GCC.hpp [hunk 2]
```diff
@@ -16,31 +16,31 @@ namespace desul {
 namespace Impl {
 
 // clang-format off
-#define DESUL_IMPL_GCC_HOST_ATOMIC_FETCH_OP_INTEGRAL_ORDER_SCOPE(OP, MEMORY_ORDER, MEMORY_SCOPE)                                 \
+#define DESUL_IMPL_GCC_HOST_ATOMIC_FETCH_OP_INTEGRAL_ORDER_SCOPE(FETCH_OP, OP_FETCH, MEMORY_ORDER, MEMORY_SCOPE)                                 \
   template <class T>                                                                                                          \
```
Comment (dalg24) [inline atomics/include/desul/atomics/Fetch_Op_GCC.hpp:20]: Same question as for hip and sycl here. Presumably this will not be in the MSVC code path.

**Review by dalg24 — COMMENTED** [review]
Presumably, only the generic fetch op needs fixing, potentially the CUDA atomics too. Why would we fix anything else if it is not broken and the new code is less readable ?

File: atomics/include/desul/atomics/Fetch_Op_Generic.hpp [hunk 3]
```diff
@@ -18,38 +18,38 @@ SPDX-License-Identifier: (BSD-3-Clause)
...
+  ANNOTATION T HOST_OR_DEVICE##_atomic_##FETCH_OP(                             \
+      T* const dest, const T val, MemoryOrder order, MemoryScope scope) {      \
+    return HOST_OR_DEVICE##_atomic_fetch_oper(                                 \
+        OP_U##operator<T, const T>(), dest, val, order, scope);                \
+  }                                                                            \
+  template <class T, class MemoryOrder, class MemoryScope>                     \
+  ANNOTATION T HOST_OR_DEVICE##_atomic_##OP_U##fetch(                          \
+      T* const dest, const T val, MemoryOrder order, MemoryScope scope) {      \
+    return HOST_OR_DEVICE##_atomic_oper_fetch(                                 \
+        OP_U##operator<T, const T>(), dest, val, order, scope);                \
```
Comment (dalg24) [inline atomics/include/desul/atomics/Fetch_Op_Generic.hpp:32]: This feels like a hack.  I would rename all the foo_operator functors to foo_fetch_operator.

**Review by dalg24 — COMMENTED** [review]
Fine by me once we confirmed Kokkos properly builds with that patch

**Comment by masterleinad** [comment]
> Fine by me once we confirmed Kokkos properly builds with that patch

https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7124/7/pipeline-console/ has two Cuda builds that are failing to run tests because of machine issues; all other CI for https://github.com/kokkos/kokkos/pull/7124 is passing.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**